Fix args for meta__fused_moving_avg_obs_fq_helper

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88058

Fixes https://github.com/pytorch/torchdynamo/issues/1802 

There are a few problems, 
1. torch.fused_moving_avg_obs_fake_quant doesn't have OpInfo test
2. self.empty_like() is not a valid call. it should be torch.empty_like(self)
3. python meta function has some unexplained behavior for arguments with default value of bool type? 

In particular, problem 3 is the most concerning one.
**UPDATE: This is expected behavior, see discussion below for explanation.** 

Without setting the default value for `per_row_fake_quant` and `symmetric_quant`, it gets the following error when running with meta tensor. 
```
meta__fused_moving_avg_obs_fq_helper() missing 2 required positional arguments: 'per_row_fake_quant' and 'symmetric_quant'
```
I can fix this by adding the default values to these two args. However, I observer something strange when examining the actual value in meta function. 

```
    print("per_row_fake_quant", per_row_fake_quant)
    print("symmetric_quant", symmetric_quant)
```

When default values are False, printed value correctly reflect the args value populated from call site. 
When default values are True, printed value is ALWAYS True, regardless of the populated value from call site. 
When default Values are None, printed value is `None` when call site set the value to 'False', printed value is 'True' when call site sets the value to 'True'. 

I also verify that this bug also affect for other meta function with default args.... 

My speculation is that this is something about pybind value packing when called from c++ dispatcher to python meta function, and default value parsing for python meta function (and other python dispatch functions) ?

I tried to find the c++ call stack, but gdb is missing symbols and C++ stacktrace is not working properly... Appreciate anyone who can point me to the source file for pybind value packing. 

cc @ezyang 
cc @bdhirsh. I know you had a fix in the symbolic shape branch... 
cc @yanboliang  who reported this bug

SherlockNoMad committed:
Fix args with default value for pyimpl kernel

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Fix args with default value for pyimpl kernel"




[ghstack-poisoned]

ezyang commented:
from WC:

<img width="783" alt="image" src="https://user-images.githubusercontent.com/13564/198911953-83ea7aa0-a080-4279-a050-d837cc240e7e.png">

SherlockNoMad commented on torch/_meta_registrations.py:
```diff
@@ -252,15 +252,18 @@ def meta__fused_moving_avg_obs_fq_helper(
+    per_row_fake_quant=False,
+    symmetric_quant=False,
 ):
+    print("per_row_fake_quant", per_row_fake_quant)
```
removed.

SherlockNoMad committed:
Update on "Fix args with default value for pyimpl kernel"


Fixes https://github.com/pytorch/torchdynamo/issues/1802 

There are a few problems, 
1. torch.fused_moving_avg_obs_fake_quant doesn't have OpInfo test
2. self.empty_like() is not a valid call. it should be torch.empty_like(self)
3. python meta function has some unexplained behavior for arguments with default value of bool type? 

In particular, problem 3 is the most concerning one.

Without setting the default value for `per_row_fake_quant` and `symmetric_quant`, it gets the following error when running with meta tensor. 
```
meta__fused_moving_avg_obs_fq_helper() missing 2 required positional arguments: 'per_row_fake_quant' and 'symmetric_quant'
```
I can fix this by adding the default values to these two args. However, I observer something strange when examining the actual value in meta function. 

```
    print("per_row_fake_quant", per_row_fake_quant)
    print("symmetric_quant", symmetric_quant)
```

When default values are False, printed value correctly reflect the args value populated from call site. 
When default values are True, printed value is ALWAYS True, regardless of the populated value from call site. 
When default Values are None, printed value is `None` when call site set the value to 'False', printed value is 'True' when call site sets the value to 'True'. 

I also verify that this bug also affect for other meta function with default args.... 

My speculation is that this is something about pybind value packing when called from c++ dispatcher to python meta function, and default value parsing for python meta function (and other python dispatch functions) ?

I tried to find the c++ call stack, but gdb is missing symbols and C++ stacktrace is not working properly... Appreciate anyone who can point me to the source file for pybind value packing. 

cc ezyang 
cc bdhirsh. I know you had a fix in the symbolic shape branch... 
cc yanboliang  who reported this bug

[ghstack-poisoned]

SherlockNoMad commented:
@pytorchbot merge -g
